### DISCUSSION TITLE
<class 'django.urls.exceptions.NoReverseMatch'>

### DISCUSSION THREAD

**Discussion by ghost** [netbox-community/netbox#17336] [Help Wanted!] [answered]
Hi,

Receiving below exception on a fresh install.
Not really much I have changed after installing apart from removing some dashboard items and adding a site.

I did some research and found out this mostly happens because of plugins but I don't have any plugins installed.

`<class 'django.urls.exceptions.NoReverseMatch'>

Reverse for 'consoleserverporttemplate_list' not found. 'consoleserverporttemplate_list' is not a valid view function or pattern name.

Python version: 3.11.2
NetBox version: 4.0.10
Plugins: None installed`

**Comment by candlerb** [comment]
Can you try searching for that string?

```
grep -R consoleserverporttemplate_list /opt/netbox
```

On my system it doesn't exist anywhere. I wonder if you have an extra spurious file lying around (although you did say it was a fresh install).

You could also try setting `DEBUG=True` in configuration.py and restarting. Then you will get a full backtrace, and that might identify the line that's having the problem.

Please also check if you are using virtualenv properly; you should not be relying on any version of Django installed globally on your system, nor any other python libraries for that matter.

**Reply by ghost** [reply]
Thanks for the reply!

Searching the file does not return any results.

I've included the debug info, maybe that helps. 
Seems to be something on the dashboard but I can't really make up much out of it.
[debug.txt](https://github.com/user-attachments/files/16846535/debug.txt)

**Reply by candlerb** [reply, marked as answer]
OK, so that comes from the ObjectCountsWidget, which does indeed appear to be broken in your setup.

In my (uncustomized) setup, I have several instances of the ObjectCountsWidget which are bound to explicit sets of models. For example, the one labelled "Organization" looks like this:

![image](https://github.com/user-attachments/assets/193ef367-5faf-4fde-957a-d657566385a0)

It would appear that you have an object counts widget configured to count ConsoleServerPortTemplates (or perhaps it's not bound to *any* model and it's trying to search *all* models??). The code is (wrongly) hard-coded to assume that every model has a "foo_list" endpoint.

You can probably fix this by resetting your dashboard.

However, I think Netbox should be made more robust against this problem. If you edit `/opt/netbox/netbox/extras/dashboard/widgets.py`
you can find:

```
class ObjectCountsWidget(DashboardWidget):
...
    def render(self, request):
        counts = []
        for model in get_models_from_content_types(self.config['models']):
            permission = get_permission_for_model(model, 'view')
            if request.user.has_perm(permission):
                url = reverse(get_viewname(model, 'list'))
                qs = model.objects.restrict(request.user, 'view')
                # Apply any specified filters
                if filters := self.config.get('filters'):
                    params = dict_to_querydict(filters)
                    filterset = getattr(resolve(url).func.view_class, 'filterset', None)
                    qs = filterset(params, qs).qs
                    url = f'{url}?{params.urlencode()}'
                object_count = qs.count
                counts.append((model, object_count, url))
            else:
                counts.append((model, None, None))

        return render_to_string(self.template_name, {
            'counts': counts,
        })
```

Then you can try changing it to (untested):

```
    def render(self, request):
        counts = []
        for model in get_models_from_content_types(self.config['models']):
            permission = get_permission_for_model(model, 'view')
            if request.user.has_perm(permission):
                try:     # <<<<< start of change
                    url = reverse(get_viewname(model, 'list'))
                except:
                    url = '#'    # <<<<< end of change
                qs = model.objects.restrict(request.user, 'view')
                # Apply any specified filters
                if filters := self.config.get('filters'):
                    params = dict_to_querydict(filters)
                    filterset = getattr(resolve(url).func.view_class, 'filterset', None)
                    qs = filterset(params, qs).qs
                    url = f'{url}?{params.urlencode()}'
                object_count = qs.count
                counts.append((model, object_count, url))
            else:
                counts.append((model, None, None))
        return render_to_string(self.template_name, {
            'counts': counts,
        })
```

(I have highlighted the changed block; one line is the existing one that has been indented by 4 extra spaces). If that works, a proper patch can be created and submitted.

IMO though, Netbox should also be more robust against broken dashboard widgets in general. If a widget raises an exception, then I think it shouldn't crash Netbox, but the homepage should still be displayed, and the widget panel should contain an error message. That would make it very obvious which widget is at fault, and also let you click the settings icon to adjust the widget to fix it.

**Reply by ghost** [reply]
Thanks! Indeed solved my problem.
Wouldn't expect the full installation to brick because of a simple dashboard widget indeed.

**Reply by candlerb** [reply]
This problem now raised as #17341 with PR #17342. EDIT: patch improved by Jeremy and merged into 4.0.11
